lib/media/streaming_engine.js

```diff
@@ -1168,6 +1194,15 @@ shaka.media.StreamingEngine = class {
           mediaState.clearBufferSafeMargin);
       return;
     }
+    // Handle deferred closeSegmentIndex() operations.
+    for (const [key, value] of this.deferredCloseSegmentIndex_.entries()) {
+      const streamId = /** @type {string} */ (key);
+      const closeSegmentIndex = /** @type {!function()} */ (value);
+      if (streamId.includes(mediaState.type)) {
+        closeSegmentIndex();
```

Comment:
Edge case: If we switch from A to B then back to A before we reach here, is this closing a distinct SegmentIndex of A, such that the currently-in-use SegmentIndex of A is still safe?

Reply:
No, thanks for pointing this out. It doesn't look like the encapsulating `Stream` objects are reinitialized on switches similar to how `SegmentIndex` is, so this would simultaneously clear out the new A and fail to clear out the old A (since we aren't storing a reference directly to the `SegmentIndex` objects, and instead the function to clear them out).

I think one way to prevent this is to store the old `SegmentIndex` objects themselves, rather than `closeSegmentIndex()`.